### PR TITLE
IBX-605: Refactored and optimized orphaned URLs deleting

### PR DESCRIPTION
**Pull request by barw4**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-605](https://issues.ibexa.co/browse/IBX-605)
| **Type**                                   | improvement
| **Target eZ Platform version** | `v3.3`
| **BC breaks**                          | no
| **Doc needed**                       | no

Narrowing down the potentially deletable URLs will result in getting rid of database deadlocks.

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 1]
```diff
@@ -198,17 +223,17 @@ public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = [])
...
     /**
-     * Delete all orphaned URLs.
+     * Delete potentially orphaned URLs.
      *
      * That could be avoided if the feature is implemented there.
      *
      * URL is orphaned if it is not linked to a content attribute through ezurl_object_link table.
      */
-    private function deleteOrphanedUrls()
+    private function deleteOrphanedUrls(array $potentiallyOrphanedUrls): void
```
Comment (adamwojs) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php:236]: `@param int[] $potentiallyOrphanedUrls` is missing

**Review by adamwojs — APPROVED** [review]

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 2]
```diff
@@ -164,19 +164,44 @@ public function linkUrl($urlId, $fieldId, $versionNo)
...
-    public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = [])
+    public function unlinkUrl($fieldId, $versionNo, array $excludeUrlIds = []): void
     {
-        $deleteQuery = $this->connection->createQueryBuilder();
+        $selectQuery = $this->connection->createQueryBuilder();
+        $selectQuery
+            ->select('link.url_id')
+            ->from($this->connection->quoteIdentifier(self::URL_LINK_TABLE), 'link')
+            ->where(
+                $selectQuery->expr()->andX(
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php:174]: Technical debt: method `andX` is deprecated.

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 3]
```diff
@@ -164,19 +164,44 @@ public function linkUrl($urlId, $fieldId, $versionNo)
...
+            return;
+        }
+
+        $deleteQuery = $this->connection->createQueryBuilder();
         $deleteQuery
             ->delete($this->connection->quoteIdentifier(self::URL_LINK_TABLE))
             ->where(
-                $deleteQuery->expr()->andX(
-                    $deleteQuery->expr()->in(
+                $selectQuery->expr()->andX(
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php:199]: Why would you use `$selectQuery` expressions inside `$deleteQuery`? The expression builder is the same, but semantically it does not make sense to me.

File: eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php [hunk 4]
```diff
@@ -164,19 +164,44 @@ public function linkUrl($urlId, $fieldId, $versionNo)
...
+                        'link.contentobject_attribute_version',
+                        ':contentobject_attribute_version'
+                    )
+                )
+            )
+            ->setParameter(':contentobject_attribute_id', $fieldId, ParameterType::INTEGER)
+            ->setParameter(':contentobject_attribute_version', $versionNo, ParameterType::INTEGER);
 
+        $statement = $selectQuery->execute();
+        $potentiallyOrphanedUrls = $statement->fetchAll(PDO::FETCH_COLUMN);
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php:189]: Technical debt:
> Method fetchAll is deprecated. Use fetchAllNumeric(), fetchAllAssociative() or fetchFirstColumn() instead.

**Review by alongosz — CHANGES_REQUESTED** [review]

**Review by alongosz — APPROVED** [review]

**Comment by piotrbisaga** [comment]
QA approved.
Check on Ibexa Commerce 3.3.6 on branch.

**Comment by lserwatka** [comment]
You can merge it up now

**Comment by adamwojs** [comment]
Done.

**Pull request merged**